File: android/app/src/main/java/com/sundeefundee/domain/WeightCalculations.kt
package com.sundeefundee.domain

import kotlin.math.roundToInt

/**
 * Result of a training session for weight progression logic.
 */
enum class SessionResult {
    FIRST,
    SUCCESS,
    FAILURE
}

/**
 * Available 1RM calculation formulas.
 */
enum class OneRepMaxFormula {
    EPLEY,
    BRZYCKI
}

/**
 * Pure functions for weight and rep calculations used in CrossFit/strength training.
 */
object WeightCalculations {

    /**
     * Rounds a weight to the nearest 5 units.
     */
    fun roundToNearestFive(value: Double): Double {
        return (value / 5.0).roundToInt() * 5.0
    }

    /**
     * Calculates target weight based on 1RM and percentage.
     */
    fun calculateTargetWeight(oneRepMax: Double, percentage: Double): Double {
        return roundToNearestFive(oneRepMax * percentage)
    }

    /**
     * Gets the next recommended weight based on session result.
     */
    fun getNextRecommendedWeight(
        currentWeight: Double,
        result: SessionResult,
        oneRepMax: Double
    ): Double {
        return when (result) {
            SessionResult.FIRST -> roundToNearestFive(oneRepMax * 0.7)
            SessionResult.SUCCESS -> roundToNearestFive(currentWeight + 5)
            SessionResult.FAILURE -> {
                val floor = roundToNearestFive(oneRepMax * 0.5)
                maxOf(roundToNearestFive(currentWeight - 5), floor)
            }
        }
    }

    /**
     * Determines if a set was successful based on actual vs prescribed values.
     */
    fun wasSetSuccessful(
        actualReps: Int,
        prescribedReps: Int,
        actualWeight: Double,
        prescribedWeight: Double?
    ): Boolean {
        val repsOk = actualReps >= prescribedReps
        val weightOk = prescribedWeight?.let { actualWeight >= it } ?: true
        return repsOk && weightOk
    }

    /**
     * Checks if a weight is a personal record compared to previous max.
     */
    fun isPersonalRecord(weight: Double, previousMax: Double?): Boolean {
        return previousMax?.let { weight > it } ?: true
    }

    /**
     * Calculates total volume load (weight × reps × sets).
     */
    fun calculateVolumeLoad(weight: Double, reps: Int, sets: Int): Double {
        return weight * reps * sets
    }

    /**
     * Detects if a plateau exists (less than 5 unit variation in last 3 weights).
     */
    fun detectPlateau(weights: List<Double>): Boolean {
        if (weights.size < 3) return false
        val last3 = weights.takeLast(3)
        return (last3.maxOrNull()!! - last3.minOrNull()!!) < 5.0
    }
}

/**
 * Epley formula for estimating 1RM.
 * Formula: weight × (1 + reps / 30)
 */
object EpleyFormula {

    /**
     * Estimates 1RM from a submaximal lift.
     * Only valid for reps > 1 (for reps == 1, the lift itself is the 1RM).
     */
    fun estimated1RM(weight: Double, reps: Int): Double {
        if (reps <= 1) return weight
        return weight * (1.0 + reps / 30.0)
    }

    /**
     * Checks if the new estimate is a PR vs the stored max.
     */
    fun isPR(newEstimate: Double, currentMax: Double?): Boolean {
        return currentMax?.let { newEstimate > it } ?: true
    }
}

/**
 * Brzycki formula for estimating 1RM.
 * Formula: weight × (36 / (37 - reps))
 */
object BrzyckiFormula {

    /**
     * Estimates 1RM using Brzycki formula.
     * Only valid for reps < 37.
     */
    fun estimated1RM(weight: Double, reps: Int): Double {
        if (reps <= 1) return weight
        if (reps >= 37) return weight * 2.0 // Approximate for very high reps
        return weight * (36.0 / (37.0 - reps))
    }
}

/**
 * Calculator for one-rep max using various formulas.
 */
object OneRepMaxCalculator {

    /**
     * Calculates 1RM using the specified formula.
     */
    fun calculate(weight: Double, reps: Int, formula: OneRepMaxFormula): Double {
        return when (formula) {
            OneRepMaxFormula.EPLEY -> EpleyFormula.estimated1RM(weight, reps)
            OneRepMaxFormula.BRZYCKI -> BrzyckiFormula.estimated1RM(weight, reps)
        }
    }

    /**
     * Calculates percentage of 1RM for a given weight.
     */
    fun calculatePercentageOf1RM(oneRepMax: Double, weight: Double): Double {
        return if (oneRepMax > 0) weight / oneRepMax else 0.0
    }

    /**
     * Gets weight for a target percentage of 1RM.
     */
    fun weightForPercentage(oneRepMax: Double, percentage: Double): Double {
        return WeightCalculations.roundToNearestFive(oneRepMax * percentage)
    }
}

File: web-app/src/lib/domain/weight-calculations.ts
import { SessionResult } from "./types";

/**
 * Rounds a value to the nearest multiple of 5.
 */
export function roundToNearestFive(value: number): number {
  return Math.round(value / 5) * 5;
}

/**
 * Calculates the target weight for a given 1RM and percentage.
 * Result is rounded to the nearest 5.
 */
export function calculateTargetWeight(oneRepMax: number, percentage: number): number {
  return roundToNearestFive(oneRepMax * percentage);
}

/**
 * Returns the next recommended weight based on previous result.
 * - first: 70% of 1RM, rounded to nearest 5
 * - success: current + 5, rounded up to nearest 5
 * - failure: current - 5, floored at 50% of 1RM (rounded to nearest 5)
 */
export function getNextRecommendedWeight(
  current: number,
  result: (typeof SessionResult)[keyof typeof SessionResult],
  oneRepMax: number
): number {
  switch (result) {
    case SessionResult.first:
      return roundToNearestFive(oneRepMax * 0.7);
    case SessionResult.success:
      return roundToNearestFive(current + 5);
    case SessionResult.failure: {
      const decreased = roundToNearestFive(current - 5);
      const floor = roundToNearestFive(oneRepMax * 0.5);
      return Math.max(decreased, floor);
    }
  }
}

/**
 * Determines whether a set was successful.
 * A set is successful if actual reps >= prescribed reps AND actual weight >= prescribed weight.
 */
export function wasSetSuccessful(
  actualReps: number,
  prescribedReps: number,
  actualWeight: number,
  prescribedWeight?: number
): boolean {
  const repsOk = actualReps >= prescribedReps;
  const weightOk = prescribedWeight === undefined ? true : actualWeight >= prescribedWeight;
  return repsOk && weightOk;
}

/**
 * Returns true if the given weight is a personal record over the previous max.
 */
export function isPersonalRecord(weight: number, previousMax: number): boolean {
  return weight > previousMax;
}

/**
 * Calculates total volume load: weight × reps × sets.
 */
export function calculateVolumeLoad(weight: number, reps: number, sets: number): number {
  return weight * reps * sets;
}

/**
 * Detects a plateau: true if the last 3 weights are all within 5 kg of each other.
 * Returns false if fewer than 3 data points.
 */
export function detectPlateau(weights: number[]): boolean {
  if (weights.length < 3) return false;
  const last3 = weights.slice(-3);
  const min = Math.min(...last3);
  const max = Math.max(...last3);
  return max - min <= 5;
}

/**
 * Estimates 1RM using the Epley formula: weight × (1 + reps / 30).
 * Returns weight directly if reps <= 1.
 */
export function estimate1RM(weight: number, reps: number): number {
  if (reps <= 1) return weight;
  return weight * (1 + reps / 30);
}

/**
 * Returns true if the new 1RM estimate exceeds the current recorded max.
 */
export function isPR(newEstimate: number, currentMax: number): boolean {
  return newEstimate > currentMax;
}
